lib/traversing.js

```diff
@@ -9,30 +9,34 @@
 function traverse(node, opt_onEnter, opt_onLeave) {
   if (opt_onEnter) opt_onEnter(node);
 
-  const childNodes = _collectChildNodes(node);
-  childNodes.forEach(function(childNode) {
-    traverse(childNode, opt_onEnter, opt_onLeave);
+  const childNodeInfo = _collectChildNodeInfo(node);
+  childNodeInfo.forEach(function([childNode, propName, parentType]) {
+    traverse(childNode, opt_onEnter ? (node, pn, pt) => {
+      opt_onEnter(node, pn || propName, pt || parentType);
+    } : null, opt_onLeave ? (node, pn, pt) => {
+      opt_onLeave(node, pn || propName, pt || parentType);
+    } : null);
   });
 
   if (opt_onLeave) opt_onLeave(node);
 }
 
 
 /**
- * @enum {number}
+ * @enum {function}
```

Comment:
never mind, each property has a slightly different type. I don't think this is the correct usage for `@enum`, because that implies that all the values have the same type. I would just get rid of it.

Admittedly they are all subtypes of Function but that's not very informative.

Reply:
SGTM